### ISSUE TITLE
About geometry transform (dtype and performance)

### ISSUE THREAD

**Issue by yxdragon** [scikit-image/scikit-image#3833]
1. why transform result come into double? Can it stay in the source's dtype? And the double is not between 0-1, what does it mean?
2. I find 'warp' transform every point in the destination image. **I want to transform many little image into a total large image, now it is very slow and cost so much memory** Can we define a mask for needed area? Or add a parameter to check the needed area? I know it is not possible to count the destination area for arbitrary function faster, but mostly, we can assume it is equal in topology. (for any point within an arbitrary polygon, after transform, the transfromed point should be also within the transfromed polygon) So we can transform the edge pix form source to destination, then fill the destination polygon as the area needed.

**Comment by hmaarrfk** [comment]
1. Can you give the code that illustrates your issue?

2. My strategy for this is to warp into smaller sub images, then to use block on the final output.

**Comment by jni** [comment]
@yxdragon 

> why transform result come into double? Can it stay in the source's dtype?

Interpolation is naturally a floating point operation if you want to operate without loss of information. We recently added the machinery needed to do float32, see #3486, but it still needs some more work to expose it to the user. It should be doable to add a dtype parameter.

> And the double is not between 0-1, what does it mean?

Cubic interpolation can result in values outside of the input range. I expect that's what's happening here.

**Comment by yxdragon** [comment]
@jni but the float result is not between 0-1, and is not the interpolation of the source image.  (My image's max value is 16000, and the transform result's max value is 0.48, what does it mean?)

And can we add a parameter to keep the type of the source image? some times it does not matter to loss little information, but we need the same type.

**Comment by yxdragon** [comment]
the second question is, I want to transform many small images into a large image (20480 * 20480), every small image would only result in a small region in the large image, but I find transform.warp count all points onf 20480^2. So if we can define a mask to indicate the area need to transfrom.

**Comment by jni** [comment]
@yxdragon 

> but the float result is not between 0-1, and is not the interpolation of the source image. (My image's max value is 16000, and the transform result's max value is 0.48, what does it mean?)

Please read http://scikit-image.org/docs/dev/user_guide/data_types.html

You should use `img_as_int16` on the result (I presume) to get back your original range and dtype.

As a side note, many users are confused by our automatic dtype and range conversions, so we are actively thinking about how to deal with it better, but without breaking everyone's existing code. See #3009 for discussion.

> the second question is, I want to transform many small images into a large image (20480 * 20480), every small image would only result in a small region in the large image, but I find transform.warp count all points onf 20480^2. So if we can define a mask to indicate the area need to transfrom.

This is a very good feature suggestion, but it doesn't exist yet. It is probably not trivial to implement. As @hmaarrfk suggests, as a temporary workaround, you could remove part of the translation component of your transform to warp the image into a smaller volume, then [montage](http://scikit-image.org/docs/dev/api/skimage.util.html#skimage.util.montage) the results together.

With regards to the feature implementation, we would also want, probably, to allow passing in an output array, so that the result is inserted there rather than returned as an enormous array (even with the mask)

**Comment by stefanv** [comment]
For affine, similarity, and projective transformations, there's a simpler solution.  You warp the four end corners (using `inverse_transform.inverse`) from the source image to the target image, which gives you the extents of the output.  You use this as your output window, and modify your transformation to take into consideration the new offset.  Then, you warp and fill in the smaller region.

**Comment by jni** [comment]
@stefanv "you use this as your output window" can you elaborate on this? I only see `output_shape` as a possible argument in the [warp docs](http://scikit-image.org/docs/dev/api/skimage.transform.html#skimage.transform.warp). Or are you just rephrasing what I said above about translations, offsets, and montaging? Also as far as I can tell, proper montaging many transforms together will be super finicky. (You need the full bounds of the transform, not just the corners?)

**Comment by stefanv** [comment]
Multiple transforms together automatically will be hard, but you can speed up transforming into a very large (mostly empty) output image by only warping *onto* a small segment of that image.  But, I think we should punt on this for now; it's not a general enough use case, and can be constructed using the tools already in place.

**Comment by yxdragon** [comment]
@stefanv yes, that is what i do now, but my transform function is geography projection with pyproj4. for linear trsmsform, the 4 point can define the result bound, and most transform has a topology equal, i mean point in a polygon should not go out after transformation, so I can transform the src image's out box pixels to get the dst bound.

but the code is not esay, especially when the dst bound go out of the image, and we must modify the transform function to adjust the small slice image, so i think that is the warp function should do.

1. support a output mask, or a vector polygon.
2. support a output buffer, and a mix mode between the transformed pixel and the dst pixel value, such as max, min, update, or mean.
3. if possible, we can support a polygon interpolate transform? geography projection is very complex, now i just transform some point, may be a grid with step 10. then use scipy.map_coordinate to generate others. i think skimage can give a help function?

**Comment by grlee77** [comment]
closed the Discussion page and moved this back here. 

One thing now outdated in the description up top is that as of `skimage` 0.19, float16 or float32 inputs will result in float32 output.
